### PR TITLE
Add `--reload` to restart tracking server when code changes for faster development

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Add `--reload` to restart tracking server when code changes for faster development.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
Test

https://user-images.githubusercontent.com/17039389/180703646-ace88aab-5cab-4c83-914c-b159ad098c07.mov

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 from mlflow.models.model import MLMODEL_FILE_NAME
 from mlflow.projects._project_spec import MLPROJECT_FILE_NAME
 from mlflow.protos import databricks_pb2
+
```
Comment (harupy) [inline mlflow/server/handlers.py:23]:
```suggestion
```

**Review by WeichenXu123 — APPROVED** [review]
LGTM

File: dev/run-dev-server.sh [hunk 2]
```diff
@@ -1,32 +1,28 @@
...
+echo 'Running tracking server in the background'
 if [ -z "$MLFLOW_TRACKING_URI" ]; then
   backend_store_uri=""
   default_artifact_root=""
 else
   backend_store_uri="--backend-store-uri $MLFLOW_TRACKING_URI"
   default_artifact_root="--default-artifact-root mlruns"
 fi
-mlflow server $backend_store_uri $default_artifact_root --gunicorn-opts="--log-level debug" > $log_file 2>&1 &
+mlflow server $backend_store_uri $default_artifact_root --gunicorn-opts="--log-level debug --reload" &
```
Comment (WeichenXu123) [inline dev/run-dev-server.sh:26]: Q: How to kill the server easily ?

Comment (harupy) [reply dev/run-dev-server.sh:26]: Ctrl + C

Comment (harupy) [reply dev/run-dev-server.sh:26]: image

Comment (WeichenXu123) [reply dev/run-dev-server.sh:26]: I think we can do:

```
spawnedChildProcess1 &
spawnedChildProcess2 &
spawnedChildProcess3 &

wait
```
in the script, so when the script process exit, all children process will also be killed.

Comment (WeichenXu123) [reply dev/run-dev-server.sh:26]: You added `&` after the `mlflow server` command, I guess ctrl+c does not work to kill it

Comment (harupy) [reply dev/run-dev-server.sh:26]: gunicorn automatically kills the child processes when the script process exits.

Comment (harupy) [reply dev/run-dev-server.sh:26]: @WeichenXu123 You can clone this PR and test it.

Comment (harupy) [reply dev/run-dev-server.sh:26]: @WeichenXu123 Did you test it?

Comment (WeichenXu123) [reply dev/run-dev-server.sh:26]: Yes. Your PR works. Interesting. gunicorn should catch parent process signal.

File: dev/run-dev-server.sh [hunk 3]
```diff
@@ -1,32 +1,28 @@
 #!/bin/bash
 set -e
 
-log_file="outputs/mlflow-server.log"
```
Comment (WeichenXu123) [inline dev/run-dev-server.sh:4]: Nit: now the mlflow server log and yarn server log will be mixed and print to screen. Is it an issue ?

Comment (harupy) [reply dev/run-dev-server.sh:4]: I don't think it is an issue.

Comment (harupy) [reply dev/run-dev-server.sh:4]: I stopped logging to a file to make it easier to check hot reloading is triggered.

Comment (harupy) [reply dev/run-dev-server.sh:4]: Having both server and yarn logs in the same terminal might be confusing, but not a big issue.

**Pull request merged**